apps/keyboard/js/imes/emoji/emoji.js

```diff
@@ -0,0 +1,121 @@
+/* global InputMethods, PromiseStorage*/
+
+
+(function () {
+  'use strict';
+
+  var KEY_COUNT_PER_PANEL = 36;
+  var EMOJI_DB_NAME = 'Emoji';
+  var EMOJI_RECENT_USED_KEY = 'emojiRecentUsedKeys';
+
+  var _glue = null;
+  var dbStore = null;
+
+  var recentKeys = {
+    _items: [],
+    _compositeBuffer: [],
+
+    get items() {
+      return this._items || [];
+    },
+    set items(val) {
+      if (Array.isArray(val) && val.length) {
+        this._items = val.slice(0, KEY_COUNT_PER_PANEL);
+      }
+    },
+
+    get panelKeys() {
+      return this._items.map((val) => {
+        return { compositeKey: val, type: 'emoji' };
+      });
+    },
+
+    add: function(keyCode) {
+      this._compositeBuffer = this._compositeBuffer.concat(keyCode);
+      if (this._compositeBuffer.length < 2) {
+        return;
+      }
+
+      var compositeKey = String.fromCharCode.apply(this, this._compositeBuffer);
+      var duplicateKeyIndex = this._items.indexOf(compositeKey);
+
+      // splice an element from array can always reduce length by 1.
+      // this ensure that we remove the last element or existent key,
+      // before unshifting clicked key into array.
+      if (this._items.length >= KEY_COUNT_PER_PANEL ||
+        duplicateKeyIndex !== -1) {
+          this._items.splice(duplicateKeyIndex, 1);
+      }
+
+      this._items.unshift(compositeKey);
+      this._compositeBuffer = [];
+    },
+    clean: function() {
+      this._items = [];
+      this._compositeBuffer = [];
+    },
+    deactivate: function () {
+      this._compositeBuffer = [];
+    }
+  };
+
+  var pageInterpolaters = {
+    recent: (page) => {
+      return new Promise((resolve) => {
+        page.panelKeys = recentKeys.panelKeys;
+
+        resolve(page);
+      });
+    }
+  };
+
+  function init(glue) {
+    _glue = glue;
+
+    dbStore = new PromiseStorage(EMOJI_DB_NAME);
+
+    var p = new Promise((resolve) => {
+      dbStore.start().then(() => {
+        dbStore.getItem(EMOJI_RECENT_USED_KEY).then((keys) => {
+          recentKeys.items = keys;
+          resolve();
+        });
+      });
+    });
+
+    return p;
+  }
+
+  function activate() {
+
+  }
+
+  function deactivate() {
+    recentKeys.deactivate();
+
+    return dbStore.setItem(EMOJI_RECENT_USED_KEY, recentKeys.items);
+  }
+
+  function click(keyCode) {
+    recentKeys.add(keyCode);
+
+    return _glue.sendKey(keyCode);
+  }
+
+
+
+  function displaysCandidates() {
+    return false;
+  }
+
+
+  InputMethods.emoji = {
+    init: init,
+    activate: activate,
+    deactivate: deactivate,
+    click: click,
+    displaysCandidates: displaysCandidates,
+    recentUsedPanelKeys: recentKeys.PanelKeys,
+    pageInterpolaters: pageInterpolaters
```

Comment:
Internally, IMEngine can either cache the promise that would resolve to the recent page keys, or the keys array itself -- a promise resolve to an object can change anyway.